jep/302/README.adoc

```diff
@@ -0,0 +1,260 @@
+The official way to revert an upgrade if something went wrong is to restore a previous backup.
+
+In the context of _Essentials_, we cannot rely on external backups to revert to the _N-1_ version: this would require some manual user intervention, which is clearly not the user experience _Essentials_ wants to provide.
+
+== Reasoning
+
+=== Why Git
+
+Using filesystem-level tools offering a snapshotting feature, like LVM, ZFS or btrfs to give a few examples, was considered.
+But this was evacuated because _Essentials_ vision is about providing an link:https://github.com/jenkinsci/jep/tree/71d9391744c8cc7d6595805f7fdd327eedf6811a/jep/300#automatically-updated-distribution["_easier to use_ and _easier to manage_ Jenkins environment"].
+As per the link:https://github.com/jenkinsci/jep/tree/71d9391744c8cc7d6595805f7fdd327eedf6811a/jep/300#target-audience[targeted audience], we obviously do not want to expect _Essentials_ users to be system experts able to set up a dedicated filesystem to operate Jenkins.
+And even with system expert, doing so would not make Essentials a very easy and quick to use distribution of Jenkins.
+
+Git offers in this matter a powerful user-space tool that allows us to version,
+and quickly rollback to some previous state if need be.
+
+Git is also a very common tool nowadays for developers,
+hence it will help making Essentials more accessible to contributors.
+
+== Why not use compatibleSinceVersion metadata
+
+TIP: A given plugin can indicate what is "the oldest version [...] configuration-compatible with.".
+In other words, if I am changing the configuration format of a plugin in non backwards compatible manner, I am _expected_ to express it as such using link:https://jenkinsci.github.io/maven-hpi-plugin/hpi-mojo.html#compatibleSinceVersion[`compatibleSinceVersion`].
+
+So, this means we could use this field to _optimize_ the way we roll back: once a configuration file has been upgraded by a plugin update, we can exclude it from rollback if the `compatibleSinceVersion` field was not changed.
+
+We decided to not do, for now at least, for two reasons:
+
+* this looks like an _optimization_. $
+Hence as such, this is probably premature to try and be very smart with the way the downgrade will work ;
+* there is no guarantee that all plugin developers currently know and think about updating this metadata when needed.
+This is likely something that will be improveable when we work on link:https://issues.jenkins-ci.org/browse/JENKINS-49806[JENKINS-49806] and write the associated Jenkins Enhancement Proposal.
+
+== Backwards Compatibility
+
+There are no backwards compatibility concerns related to this proposal.
+
+== Security
+
+=== Secrets
+
+Versioning secrets is not an issue, as the snapshotting system is going to be local to the running instance.
+As the Git repository data will never be pushed _outside_, no data leak is expected from this side.
```

Comment:
I do not think this is acceptable. Someone will use the Git repository externally regardless of what you advertise.

Fortunately there is a simple solution: add

```
secrets/master.key
```

to `.gitignore`. It is created on first startup and never changed thereafter, so it would never be affected by an upgrade or downgrade. And without it, all other secrets remain (theoretically) secret.

[Reference](https://go.cloudbees.com/docs/cloudbees-documentation/cje-user-guide/index.html#backup-sect-configure)

Reply:
Well, I do not feel strongly to add it or ignore it. But well, if someone is able to access the FS and so what you are saying, in any case they have the power to read that key.

I understand what you are saying as "someone could have the bad idea to push that repo elsewhere, when it's not supposed to be, not groking the fact they're going to leak central secrets doing so, and so we'll be seen as responsible for it.".

So yes, let's be conservative and not encourage people to shoot themselves in the foot.

Reply:
FTR, fixed.